Pull request issue 247

zachsa opened the pull request:
1. Updated Segment API to allow for programmatic segment property updates
2. Updated test suite to cover validated, programmatic segment updates

zachsa commented:
This pull request addresses issue #247

zachsa commented on src/main/markers/waveform.segments.js:
```diff
@@ -126,32 +126,8 @@ define([
-
     var segment = new Segment(
+      this,
       Utils.isNullOrUndefined(options.id) ? this._getNextSegmentId() : options.id,
```
This could also be moved into segment.js since segments now reference parent objects. Then the waveform-segments object doesn't need to reference Utils at all.

chrisn commented on src/main/views/segments-layer.js:
```diff
@@ -48,6 +48,14 @@ define([
   SegmentsLayer.prototype._registerEventHandlers = function() {
     var self = this;
 
+    this._peaks.on('segments.update', function(segment) {
```
We'll need to also update the UI objects if the `startTime` or `endTime` changes. This is a bit more complex than changing the label, as a change to the start/end times could affect whether a segment remains (or becomes) visible in the waveform view.

zachsa replied:
Looking at the event handler for the `segments.add` event, I see this code:

```
this._peaks.on('segments.add', function(segments) {
  var frameOffset = self._view.getFrameOffset();
  var width = self._view.getWidth();

  var frameStartTime = self._view.pixelsToTime(frameOffset);
  var frameEndTime   = self._view.pixelsToTime(frameOffset + width);

  segments.forEach(function(segment) {
    if (segment.isVisible(frameStartTime, frameEndTime)) {
      self._addSegmentGroup(segment);
    }
  });

  self.updateSegments(frameStartTime, frameEndTime);
});
```

It looks like I should just include the same logic in the `segments.updated` event handler?

zachsa replied:
i.e... the value of `isVisible` needs to be reset if necessary, and then `self.updateSegments` should be called with new frameStartTime, frameEndTime?

chrisn replied:
Because `segments.updated` will only receive one segment, we only need to check that particular segment, rather than all of them, so I would avoid calling `updateSegments`. I guess a way to implement would be to remove the segment if it's currently visible, then add it in its new position, maybe something like this:

```javascript
var redraw = false;

var segmentGroup = this._segmentGroups[segment.id];

if (segmentGroup) {
  this._removeSegment(segment);
  redraw = true;
}

var frameOffset = self._view.getFrameOffset();
var width = self._view.getWidth();

var frameStartTime = self._view.pixelsToTime(frameOffset);
var frameEndTime   = self._view.pixelsToTime(frameOffset + width);

if (segment.isVisible(frameStartTime, frameEndTime)) {
  this._addSegmentGroup(segment);
  redraw = true;
}

if (redraw) {
  this._layer.draw();
}
```

chrisn commented on test/unit/segment-spec.js:
```diff
@@ -1,59 +1,141 @@
+  });
+
+  describe('startTime', function() {
+    it('should be possible to update programatically', function() {
```
I just realised an issue with changing the start and end times. If you have a segment, for example at (5, 10) and you want to move it to (15, 20), you have to update the end time first, otherwise you get an exception. I wonder if we should prefer to have an `updateTime(start, end)` method to be able to set both together? Or maybe to flip the start and end times automatically if you set the end time after the start time?

zachsa replied:
ah okay. I think it would make sense if the startTime/endTime were always updated together in this case. So NOT providing a means of updating starTime and endTime, and instead providing a setter function `updateTime(start, end)`?

zachsa replied:
This should be resolved now

chrisn replied:
I wonder if the API would be simpler if we added a `segment.update()` method, that takes an object with the values you want to change, rather than use getter and setters. For example:

```javascript
segment.update({ labelText: 'New label' });
segment.update({ labelText, 'New label', startTime: 20, endTime: 30 });
```

What do you think?

zachsa replied:
I agree that is a better approach

zachsa replied:
This is added in commit #6f32044

chrisn replied:
Ah yes, we don't currently use any ES6 features... although I would like to modernise the code at some point.

zachsa commented:
@chrisn - I think I have addressed everything you raised. Let me know if you have any more comments. I also added a `Segment API` section to the documentation. You can view this in my forked repository here: https://github.com/zachsa/peaks.js

chrisn commented:
Many thanks, this looks really good. I'll do some tests. Would you mind including the documentation changes in the pull request?

The only other thing I would like to do is modify the Points API in the same way, for consistency. I'm happy to do that, if you prefer?

zachsa commented:
I'll have a look at the Point API - should be pretty quick. The documentation changes are already in the pull request.

chrisn commented:
Thanks! One final task is to update the TypeScript declarations file with the API changes: https://github.com/bbc/peaks.js/blob/master/peaks.js.d.ts

zachsa commented:
> Many thanks, this looks really good. I'll do some tests. Would you mind including the documentation changes in the pull request?
> 
> The only other thing I would like to do is modify the Points API in the same way, for consistency. I'm happy to do that, if you prefer?

I'm struggling to get the points to redraw correctly using the same approach as is used for the segments. This doesn't work:

```
    this._peaks.on('points.update', function(point) {
      var redraw = false;
      var pointGroup = self._pointGroups[point.id];
      var frameOffset = self._view.getFrameOffset();
      var width = self._view.getWidth();
      var frameStartTime = self._view.pixelsToTime(frameOffset);
      var frameEndTime   = self._view.pixelsToTime(frameOffset + width);

      if (pointGroup) {
        self._removePoint(point);
        redraw = true;
      }

      if (point.isVisible(frameStartTime, frameEndTime)) {
        self._addPointGroup(point);
        redraw = true;
      }

      if (redraw) {
        self._layer.draw();
      }
    });
```

With this code, I have to 'jiggle' the waveform to get the point to display.

this code works:

```
    this._peaks.on('points.update', function(point) {
      var frameOffset = self._view.getFrameOffset();
      var width = self._view.getWidth();
      var frameStartTime = self._view.pixelsToTime(frameOffset);
      var frameEndTime   = self._view.pixelsToTime(frameOffset + width);

      self._removePoint(point);
      if (point.isVisible(frameStartTime, frameEndTime)) {
        self._addPointGroup(point);
      }

      self.updatePoints(frameStartTime, frameEndTime);
    });
```

Is it okay?

zachsa commented:
Point API is adjusted to match the Segment API. Also the updated the test suite and the documentation, similarly to the changes for the segment API. The readme file is viewable on my forked repository in master branch (and is included in the pull-request for this repo).

(https://github.com/zachsa/peaks.js#point-api)

zachsa commented:
> Thanks! One final task is to update the TypeScript declarations file with the API changes: https://github.com/bbc/peaks.js/blob/master/peaks.js.d.ts

I think I have done this... Please let me know if I haven't done this correctly as I haven't really worked with typescript or typescript definitions before.

chrisn commented on peaks.js.d.ts:
```diff
@@ -11,6 +11,7 @@ declare module 'peaks.js' {
     color?: string;
     labelText?: string;
     id?: string;
+    update: (options: Object) => void;
```
Using object is proabably too generic. I'd suggest:
```typescript
update: (options: SegmentUpdateValues) => void;
```
and
```typescript
interface SegmentUpdateValues {
  startTime?: number;
  endTime?: number;
  editable?: boolean;
  color?: string;
  labelText?: string;
}
```
(and similar for `Point.update`)

zachsa replied:
done

chrisn commented:
All good, so I have merged to `master`. Thank you for all your work on this!

zachsa commented:
Thank you!
